fel/NSFDSolverWithParticles.h

```diff
@@ -0,0 +1,382 @@
+                gamma_beta);  // Gamma-beta attribute (product of relativistiv gamma and beta)
+            this->addAttribute(R_np1);     // Position attribute for the next time step
+            this->addAttribute(R_nm1);     // Position attribute for the next time step
+            this->addAttribute(E_gather);  // Electric field attribute for particle gathering
+            this->addAttribute(B_gather);  // Magnedit field attribute for particle gathering
+        }
+
+        // Destructor for the Bunch class
+        ~Bunch() {}
+
+        // Define container types for various attributes
+        using charge_container_type   = ippl::ParticleAttrib<scalar>;
+        using velocity_container_type = ippl::ParticleAttrib<ippl::Vector<scalar, 3>>;
+        using vector_container_type   = ippl::ParticleAttrib<ippl::Vector<scalar, 3>>;
+        using vector4_container_type  = ippl::ParticleAttrib<ippl::Vector<scalar, 4>>;
+
+        // Declare instances of the attribute containers
+        charge_container_type Q;             // Charge container
+        charge_container_type mass;          // Mass container
+        velocity_container_type gamma_beta;  // Gamma-beta container
+        typename ippl::ParticleBase<PLayout>::particle_position_type
+            R_np1;  // Position container for the next time step
+        typename ippl::ParticleBase<PLayout>::particle_position_type
+            R_nm1;  // Position container for the previous time step
+        ippl::ParticleAttrib<ippl::Vector<scalar, 3>>
+            E_gather;  // Electric field container for particle gathering
+        ippl::ParticleAttrib<ippl::Vector<scalar, 3>>
+            B_gather;  // Magnetic field container for particle gathering
+    };
+    template <typename T>
+    KOKKOS_INLINE_FUNCTION Kokkos::pair<ippl::Vector<int, 3>, ippl::Vector<T, 3>> gridCoordinatesOf(
+        const ippl::Vector<T, 3> hr, const ippl::Vector<T, 3> origin, ippl::Vector<T, 3> pos) {
+        // return pear<ippl::Vector<int, 3>, ippl::Vector<T, 3>>{ippl::Vector<int, 3>{5,5,5},
+        // ippl::Vector<T, 3>{0,0,0}}; printf("%.10e, %.10e, %.10e\n", (inverse_spacing *
+        // spacing)[0], (inverse_spacing * spacing)[1], (inverse_spacing * spacing)[2]);
+        Kokkos::pair<ippl::Vector<int, 3>, ippl::Vector<T, 3>> ret;
+        // pos -= spacing * T(0.5);
+        ippl::Vector<T, 3> relpos  = pos - origin;
+        ippl::Vector<T, 3> gridpos = relpos / hr;
+        ippl::Vector<int, 3> ipos;
+        ipos = gridpos.template cast<int>();
+        ippl::Vector<T, 3> fracpos;
+        for (unsigned k = 0; k < 3; k++) {
+            fracpos[k] = gridpos[k] - (int)ipos[k];
+        }
+        // TODO: NGHOST!!!!!!!
+        ipos += ippl::Vector<int, 3>(1);
+        ret.first  = ipos;
+        ret.second = fracpos;
+        return ret;
+    }
+    template <typename view_type, typename scalar>
+    KOKKOS_FUNCTION void scatterToGrid(const ippl::NDIndex<3>& ldom, view_type& view,
```

Comment:
comment from previous review still applies: the scatter is repeated here and also in the FreeElectronLaser.cpp, would be nice to avoid this by putting it in some common file maybe.